File: 2021/python/day16.py
def load_file():
    with open('../inputs/input16') as file:
        return file.read().strip()


def bin_repr(value, base=2):
    return '{:04b}'.format(int(value, base))


def to_binary(data):
    out = ''
    for c in data:
        out += bin_repr(c, 16).zfill(4)
    out = out
    return out


class Packet:
    def __init__(self):
        self.literals = []
        self.version = 0
        self.type_id = 0
        self.packets = []


def read(data):
    packet = Packet()
    packet.version, data = int(data[:3], 2), data[3:]
    packet.type_id, data = int(data[:3], 2), data[3:]
    if packet.type_id == 4:
        literal = ''
        while data[0] != '0':
            literal, data = literal + bin_repr(data[1:5]), data[5:]
        packet.literals.append(int(literal + bin_repr(data[1:5]), 2))
        data = data[5:]
    else:
        length_type_id, data = data[0], data[1:]
        if length_type_id == '0':
            total_length_in_bits, data = int(data[:15], 2), data[15:]
            bit_count = len(data)
            while bit_count - len(data) < total_length_in_bits:
                subpacket, data = read(data)
                packet.packets.append(subpacket)
        else:
            number_of_packets, data = int(data[:11], 2), data[11:]
            for _ in range(number_of_packets):
                subpacket, data = read(data)
                packet.packets.append(subpacket)
        if packet.type_id == 0:
            s = 0
            for p in packet.packets:
                s += sum(p.literals)
            packet.literals.append(s)
        elif packet.type_id == 1:
            s = 1
            for p in packet.packets:
                for literal in p.literals:
                    s *= literal
            packet.literals.append(s)
        elif packet.type_id == 2:
            literals = []
            for p in packet.packets:
                literals.extend(p.literals)
            packet.literals.append(min(literals))
        elif packet.type_id == 3:
            literals = []
            for p in packet.packets:
                literals.extend(p.literals)
            packet.literals.append(max(literals))
        elif packet.type_id == 5:
            assert len(packet.packets) == 2
            a = packet.packets[0].literals[0]
            b = packet.packets[1].literals[0]
            packet.literals.append(int(a > b))
        elif packet.type_id == 6:
            assert len(packet.packets) == 2
            a = packet.packets[0].literals[0]
            b = packet.packets[1].literals[0]
            packet.literals.append(int(a < b))
        elif packet.type_id == 7:
            assert len(packet.packets) == 2
            a = packet.packets[0].literals[0]
            b = packet.packets[1].literals[0]
            packet.literals.append(int(a == b))
    return packet, data


def sum_versions(packet):
    s = packet.version
    for p in packet.packets:
        s += sum_versions(p)
    return s


def part1():
    data = to_binary(load_file())
    packet, _ = read(data)

    sum_version_numbers = packet.version
    for p in packet.packets:
        sum_version_numbers += sum_versions(p)
    return sum_version_numbers


def part2():
    data = to_binary(load_file())
    data = to_binary("9C0141080250320F1802104A08")
    packet, _ = read(data)
    return packet.literals[0]


print(part1())
print(part2())

File: 2021/cpp/day_16.cpp
//
// Created by trox667 on 16.12.21.
//

#include "utils.hpp"
#include "day_16.hpp"
#include <bitset>
#include <limits>


namespace day16 {
    struct Packet {
        int version;
        int type_id;
        std::vector<long> literals;
        std::vector<Packet> subpackets;

        [[nodiscard]] long sum_literals() {
            auto s = 0L;
            for (const auto &sp: subpackets) {
                for (const auto &lit: sp.literals) {
                    s += lit;
                }
            }
            return s;
        }

        [[nodiscard]] long mul_literals() {
            auto s = 1L;
            for (const auto &sp: subpackets) {
                for (const auto &lit: sp.literals) {
                    s *= lit;
                }
            }
            return s;
        }

        [[nodiscard]] long min_literal() {
            auto m = std::numeric_limits<long>::max();
            for (const auto &sp: subpackets) {
                for (const auto &lit: sp.literals) {
                    if (lit < m) m = lit;
                }
            }
            return m;
        }

        [[nodiscard]] long max_literal() {
            auto m = std::numeric_limits<long>::min();
            for (const auto &sp: subpackets) {
                for (const auto &lit: sp.literals) {
                    if (lit > m) m = lit;
                }
            }
            return m;
        }

        [[nodiscard]] long greater() {
            return subpackets[0].literals[0] > subpackets[1].literals[0] ? 1L
                                                                         : 0L;
        }

        [[nodiscard]] long less() {
            return subpackets[0].literals[0] < subpackets[1].literals[0] ? 1L
                                                                         : 0L;
        }


        [[nodiscard]] long eq() {
            return subpackets[0].literals[0] == subpackets[1].literals[0] ? 1L
                                                                          : 0L;
        }
    };

    [[nodiscard]] std::string
    take_n(int &index, int size, const std::string &data) {
        auto s = data.substr(index, size);
        index += size;
        return s;
    }

    [[nodiscard]] Packet read(int &index, const std::string &data) {
        auto packet = Packet{};
        packet.version = static_cast<int>(utils::int_from_binary_str(
                take_n(index, 3, data)));
        packet.type_id = static_cast<int>(utils::int_from_binary_str(
                take_n(index, 3, data)));
        if (packet.type_id == 4) {
            auto literal = std::string{};
            while (data.at(index) != '0') {
                index++;
                literal += take_n(index, 4, data);
            }
            index++;
            literal += take_n(index, 4, data);
            packet.literals.emplace_back(utils::int_from_binary_str(literal));
        } else {
            auto length_type_id = utils::int_from_binary_str(
                    take_n(index, 1, data));
            if (length_type_id == 0) {
                auto total_length = utils::int_from_binary_str(
                        take_n(index, 15, data));
                auto bit_count = data.size() - index;
                while (bit_count - (data.size() - index) < total_length) {
                    auto subpacket = read(index, data);
                    packet.subpackets.push_back(subpacket);
                }
            } else {
                auto number_of_packets = utils::int_from_binary_str(
                        take_n(index, 11, data));
                for (auto i = 0; i < number_of_packets; ++i) {
                    auto subpacket = read(index, data);
                    packet.subpackets.push_back(subpacket);
                }
            }
            switch (packet.type_id) {
                case 0:
                    packet.literals.emplace_back(packet.sum_literals());
                    break;
                case 1:
                    packet.literals.emplace_back(packet.mul_literals());
                    break;
                case 2:
                    packet.literals.emplace_back(packet.min_literal());
                    break;
                case 3:
                    packet.literals.emplace_back(packet.max_literal());
                    break;
                case 5:
                    packet.literals.emplace_back(packet.greater());
                    break;
                case 6:
                    packet.literals.emplace_back(packet.less());
                    break;
                case 7:
                    packet.literals.emplace_back(packet.eq());
                    break;
            }
        }

        return packet;
    }

    [[nodiscard]] int sum_version(const Packet &packet) {
        auto sum = packet.version;
        for (const auto &p: packet.subpackets) {
            sum += sum_version(p);
        }
        return sum;
    }

    [[nodiscard]] long part1(const std::vector<std::string> &lines) {
        auto data = utils::to_binary(utils::trim(lines[0]));
        auto index = 0;
        auto packet = read(index, data);
        return sum_version(packet);
    }

    [[nodiscard]] long part2(const std::vector<std::string> &lines) {
        auto data = utils::to_binary(utils::trim(lines[0]));
        auto index = 0;
        auto packet = read(index, data);
        return !packet.literals.empty() ? packet.literals[0] : 0;
    }
}
